README.md

```diff
@@ -226,15 +228,21 @@ docker build -t s3-resource .
 #### Integration tests
 
 The integration requires two AWS S3 buckets, one without versioning and another
-with. The `docker build` step requires setting `--build-args` so the
-integration will run.
+with versioning enabled. The `docker build` step requires setting `--build-args`
+so the integration will run.
 
 Run the tests with the following command:
 
 ```sh
 docker build . -t s3-resource --build-arg S3_TESTING_ACCESS_KEY_ID="access-key" --build-arg S3_TESTING_SECRET_ACCESS_KEY="some-secret" --build-arg S3_TESTING_BUCKET="bucket-non-versioned" --build-arg S3_VERSIONED_TESTING_BUCKET="bucket-versioned" --build-arg S3_TESTING_REGION="us-east-1" --build-arg S3_ENDPOINT="https://s3.amazonaws.com"
 ```
 
+### Notes
+
+The resource will cache uploaded (`put`) and downloaded (`get`) files. If
```

Comment:
Ahh, now the note above kind of makes sense. So basically if you're using `regexp` *and* a versioned S3 bucket *and* update one of the versioned files in-place (instead of giving it a new version), *then* the cache will not be what you expect.

I think this is the case regardless of whether the bucket is versioned, though. Really, this is a property of caching with *any* resource - if an external object changes out-of-band and the same "version" now has different content, the cache will be stale. Concourse expects all versions to be immutable, so this is no surprise to me, though maybe it belongs somewhere in general Concourse docs so that each and every resource type doesn't need to have a note to this effect.

In any case, I could see a feature being added to this resource where if you *are* using a versioned s3 bucket in combination with `regexp:`, it could include the versioned object identifier in the resource versions in order to detect when the file has changed in-place, thereby forcing new versions and busting the cache. Old versions would still fetch the old file (using bucket versioning semantics), but it would at least be more explicit what's happening.

Reply:
Actually, was told by the Toolsmith team that versioned files are always checked, not so for regexp files. Correct me if that is wrong.

Reply:
Not sure what you mean by always checked?

Versioned files fundamentally don't have the caching problem because they're actually immutable, and their Concourse versioning and S3 bucket versioning encapsulates that.

Regexp-versioned files have this problem because it's possible for someone to update the same filename in-place, while keeping the "version" the same (because it's just part of the filename).

When configured with `regexp:` and the S3 bucket uses object versioning, the caching problem would be fixed by the resource using both the filename *and* the S3 object version in the Concourse versions.

Reply:
If you use versioned files, it always pulls the latest version - or so I was told. So caching is not an issue. It might not do so if you use regexp and Concourse thinks it has a reasonable up-to-date file in the cache.

Then again, that info might be wrong, I'm not that deep into the details.

The confusing part here is that if you mix both (versioned files and regexp), you are up for trouble, and that should be avoided. I'm open for better wording though.

Reply:
(:turtle: sorry this is so late)

It doesn't pull the latest version, it works because `check` is able to actually detect the new version - when configured with `versioned_file:`, the resource emits and fetches versions based on S3's reported object version, not the filename.

The behavior of `regexp:` is currently unaffected by whether the bucket has versioning enabled. The fundamental issue is that the versioning is by filename. Regardless of whether the S3 bucket is versioned, Concourse is going to cache by the only information the S3 resource currently puts in the version: the filename.

I don't think this is a risky combination or anything to do with versioned files - the problem comes up if your workflow involves updating files in-place and you've pointed an S3 resource at it configured with `regexp:`. I'm OK with documenting this pitfall, I just want to make sure the README is clear about the situation so as to prevent misconceptions.